### ISSUE TITLE
Cannot run a script using `cabal run` which depends on a local package

### ISSUE THREAD

**Issue by LaurentRDC** [haskell/cabal#8562]
# Description

Let's say I have a project with an executable and a library. The repository looks like:

```txt
app/
   app/Main.hs
lib/
   lib/MyLib.hs
scripts/
   scripts/MyScript.hs
my-lib.cabal
```

I'd like to run a script (`scripts/MyScript`) which looks like this:

```haskell
import qualified MyLib

main = pure ()
```

The documentation for [`cabal run`](https://cabal.readthedocs.io/en/stable/cabal-commands.html#cabal-run) shows how to include dependencies in the script file:

```haskell
#!/usr/bin/env cabal
{- cabal:
build-depends: base
             , my-lib
-}
import qualified MyLib

main = pure ()
```

The command `cabal run scripts/MyScript.hs` with cabal-install 3.6.2.0 works as expected.However, with cabal=install 3.8.1.0 I get an error like this:

```shell
Error: cabal: Could not resolve dependencies:
[__0] trying: fake-package-0 (user goal)
[__1] unknown package: my-lib (dependency of fake-package)
[__1] fail (backjumping, conflict set: fake-package, my-lib)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: fake-package, my-lib
```

# To reproduce

[I have set up a repository to show this behavior](https://github.com/LaurentRDC/cabal-3.8-script-issue):
```
$ git clone https://github.com/LaurentRDC/cabal-3.8-script-issue && cd cabal-3.8-script-issue
$ ghcup set cabal 3.6.2.0
$ cabal v2-run scripts/MyScript.hs
$ ghcup set cabal 3.8.1.0
$ cabal v2-run scripts/MyScript.hs
```

# Expected behavior

I'm expecting to be able to compile and run `scripts/MyScript.hs`

# System information
 - Ubuntu 18.04
 - `cabal-install` version 3.8.1.0, `ghc` 9.4.2

**Additional context**

**Comment by gbaz** [comment]
I suspect that the regression is likely due to refactors induced by https://github.com/haskell/cabal/pull/7851

**Comment by ulysses4ever** [comment]
Yes, let’s page in @bacchanalia.

**Comment by bacchanalia** [comment]
I'll try to take a look at it next week.

**Comment by dschrempf** [comment]
I also have this problem. I tried adding a `cabal.project` file referencing the local package (using `packages: ..`) to no avail.

**Comment by LaurentRDC** [comment]
@bacchanalia if you don't have time to take a look, would you be able to orient me so I can see what's going on? I.e. which Cabal module is responsible for determining if local packages are in-scope?

Unfortunately, all commits from #7851 were squashed, so I can't use git-bisect to pinpoint where exactly is the regression

**Comment by Mikolaj** [comment]
@LaurentRDC: the commits seem preserved in the PR, so presumably also on the [bacchanalia:master](https://github.com/bacchanalia/cabal) branch. Is it any help?

**Comment by ulysses4ever** [comment]
I understand the desire to get back the feature that suddenly disappeared. But when I think about it in isolation of cabal history, it feels to me that scripts are meant to be self contained, and I consider them using a cabal.project a bug rather than a feature. Scripts support (a part of) cabal.project fields inside themselves not for nothing: for me, it's a clear indication that you're not supposed to have an external cabal.project.

Overall, this ticket sounds to me like a feature request to extend (back) abilities of scripts beyond what they are meant to be. It's a fine request, of course, but if not for the history, I'd probably prefer this ability to be controlled by a flag, which, by default, would ignore nearby project files. But then, there's "if not for the history bit", so perhaps it should be the other way around.

Another thought: if you really need a setup sophisticated enough to reach beyond basic project file capabilities supported in the	 script header, then you might as well bite the bullet and turn your code into a proper package. Sometimes, people dislike creating .cabal files, but with `cabal init` this is very-very easy. Maybe you'd want to give it a try while waiting for this feature request to be acted upon.

**Comment by fgaz** [comment]
I agree with @ulysses4ever. You could add a cabal.project to your project, with something like `packages: my-lib.cabal scripts/scripts.cabal`, and have a `scripts` package under the `scripts` directory with one executable per script. then you can `cabal run scripts:myscript`.
I think cabal scripts are mostly useful _outside_ of projects

**Comment by dschrempf** [comment]
In this case, I think it would be useful to specify specific versions/revisions of script dependencies (i.e., versions not available on Hackage but only on Gitlab or GitHub, or somewhere else). Is this possible in a cabal script?

**Comment by LaurentRDC** [comment]
> agree with @ulysses4ever. You could add a cabal.project to your project, with something like `packages: my-lib.cabal scripts/scripts.cabal`, and have a `scripts` package under the `scripts` directory with one executable per script. then you can `cabal run scripts:myscript`.
> I think cabal scripts are mostly useful _outside_ of projects

Thank you @fgaz , that's a great suggestion which works well. I'm satisfied.

**Comment by fgaz** [comment]
@dschrempf see #8024. `packages` should work as long as you provide absolute paths, and it looks like `source-repository-package` doesn't work but I think it should.

**Comment by ulysses4ever** [comment]
So, is resolution here: won't fix because doesn't align with the idea of scripts? It may still be a reasonable feature request...
